.changeset/ninety-eggs-argue.md

```diff
@@ -0,0 +1,79 @@
+---
+'@backstage/plugin-search-backend-node': minor
+'@backstage/create-app': patch
+---
+
+**BREAKING**: `IndexBuilder.addCollator()` now requires a `schedule` parameter (replacing `defaultRefreshIntervalSeconds`) which is expected to be a `TaskRunner` that is configured with the desired search indexing schedule for the given collator.
+
+`Scheduler.addToSchedule()` now takes a new parameter object (`ScheduleTaskParameters`) with two new options `id` and `scheduledRunner` in addition to the migrated `task` argument.
+
+NOTE: The search backend plugin now creates a dedicated database for coordinating indexing tasks.
+
+To make this change to an existing app, make the following changes to `packages/backend/src/plugins/search.ts`:
+
+```diff
++import { Duration } from 'luxon';
+
+/* ... */
+
++  const schedule = env.scheduler.createScheduledTaskRunner({
++    frequency: Duration.fromObject({ seconds: 600 }),
++    timeout: Duration.fromObject({ seconds: 900 }),
++  });
+
+   indexBuilder.addCollator({
+-    defaultRefreshIntervalSeconds: 600,
++    schedule,
+     factory: DefaultCatalogCollatorFactory.fromConfig(env.config, {
+      discovery: env.discovery,
+      tokenManager: env.tokenManager,
+     }),
+   });
+
+   indexBuilder.addCollator({
+-    defaultRefreshIntervalSeconds: 600,
++    schedule,
+     factory: DefaultTechDocsCollatorFactory.fromConfig(env.config, {
+      discovery: env.discovery,
+      tokenManager: env.tokenManager,
+     }),
+   });
+
+/* ... */
+```
+
+NOTE: For scenarios where the `lunr` search engine is used in a multi-node configuration, a non-distributed `TaskRunner` like the following should be implemented to ensure consistency across nodes:
```

Comment:
Does the default task runner work on a sqlite connection?  If so, maybe a simpler recommendation would be to have integrators configure the search plugin to use sqlite as its database instead. [e.g.](https://backstage.io/docs/tutorials/configuring-plugin-databases#postgresql-and-sqlite-3)

Reply:
And I still wonder if it's possible to detect/warn about this somewhere in the index builder.  ...But I wouldn't call that a blocker.

Reply:
AFAIK, the default task runner _should_ be DB agnostic. @freben any limitations you know of?

Reply:
It certainly works locally for me, at least!  Could simplify things quite a bit / de-emphasize the use-case with instructions like...

````md
NOTE: Although the following scenario is not recommended, if you run the in-memory `lunr` search engine in a multi-node configuration, you must also configure the `search` plugin to use an in-memory database in order to ensure index consistency across nodes:

```diff
backend:
  database:
    client: pg
    connection: 'postgresql://foo:bar@some.example-pg-instance.tld:5432'
+    plugin:
+      search:
+        client: better-sqlite3
+        connection: ':memory:'
```
````

Reply:
I can fix the documentation in a follow-up PR (I wanna spend a little bit of time to updating the architecture diagram to better highlight this anyway).